sway-types/src/span.rs

```diff
@@ -11,59 +11,27 @@ lazy_static! {
     static ref DUMMY_SPAN: Span = Span::new(Arc::from(""), 0, 0, None).unwrap();
 }
 
-pub struct Position {
-    input: Arc<str>,
+pub struct Position<'a> {
+    input: &'a str,
     pos: usize,
 }
 
-impl Position {
-    pub fn new(input: Arc<str>, pos: usize) -> Option<Position> {
-        input.clone().get(pos..).map(|_| Position { input, pos })
+impl<'a> Position<'a> {
+    pub fn new(input: &'a str, pos: usize) -> Option<Position<'a>> {
+        input.get(pos..).map(|_| Position { input, pos })
     }
+
     #[inline]
     pub fn line_col(&self) -> (usize, usize) {
         if self.pos > self.input.len() {
             panic!("position out of bounds");
         }
 
-        let mut pos = self.pos;
-        // Position's pos is always a UTF-8 border.
-        let slice = &self.input[..pos];
-        let mut chars = slice.chars().peekable();
-
-        let mut line_col = (1, 1);
-
-        while pos != 0 {
-            match chars.next() {
-                Some('\r') => {
-                    if let Some(&'\n') = chars.peek() {
-                        chars.next();
-
-                        if pos == 1 {
-                            pos -= 1;
-                        } else {
-                            pos -= 2;
-                        }
-
-                        line_col = (line_col.0 + 1, 1);
-                    } else {
-                        pos -= 1;
-                        line_col = (line_col.0, line_col.1 + 1);
-                    }
-                }
-                Some('\n') => {
-                    pos -= 1;
-                    line_col = (line_col.0 + 1, 1);
-                }
-                Some(c) => {
-                    pos -= c.len_utf8();
-                    line_col = (line_col.0, line_col.1 + 1);
-                }
-                None => unreachable!(),
-            }
-        }
-
-        line_col
+        let slice = &self.input[..self.pos];
```

Comment:
This is nice, but I think you can avoid the `collect` here.

```rust
#[inline]
    pub fn line_col(&self) -> (usize, usize) {
        if self.pos > self.input.len() {
            panic!("position out of bounds");
        }

        let (line_count, last_line) = self.input[..self.pos]
            .lines()
            .fold((0, None), |(line_count, _), line| {
                (line_count + 1, Some(line))
            });
        let last_line_len = last_line.map(|x| x.chars().count()).unwrap_or_default();
        (line_count, last_line_len)
    }
```

Reply:
+1 on usage of `.lines()` instead of splitting with `\n` if there are no drawbacks to do so. We do not officially support Windows but if somehow we decide to support offloading handling different line endings to `std` sounds nice :)

Reply:
Thanks for the suggestion @xunilrj. This version breaks all the LSP tests unfortunately and the numbers seem to be off by quite a bit. Perhaps we can look at modifying this function in a follow up PR..